### ISSUE TITLE
isolate/context separation (security conserns)

### ISSUE THREAD

**Issue by pkit** [plv8/plv8js#336]
Currently plv8 jails different users into different contexts which is fine.
But there are some security problems with that, namely a buggy user code can exhaust all memory on the server and it will lead to v8 isolate crash/exit which will kill all other contexts.
A better approach would be to use 1 isolate with 1 context per user, unused isolates do not consume more resources than unused contexts.
Memory usage can be controlled through intermittent `GetHeapStatistics()` and a custom `ArrayBuffer` allocator. Are you interested in PR(s) for that?

**Comment by JerrySievert** [comment]
I would be interested in PRs for these, yes, but I will note that since pg forks for each connection, I haven't prioritized this higher - the isolate goes with the fork.  unless v8 is doing some shared process memory magic behind the scenes that I hadn't seen before, it should become a separate copy at fork() time.  this still doesn't stop the usage of memory by a badly behaving user, but since queries/connections do not run in parallel on a single process, when that users javascript is complete, the context should go away.

that doesn't stop the possibility of two roles being used serial in a single long-lived connection (due to pooling), but I haven't seen a case when memory exhaustion hasn't also killed the forked pg process.

thoughts?

**Comment by JerrySievert** [comment]
btw, I will note that I am a fan of zerovm :)

**Comment by pkit** [comment]
The latter case is exactly the case I have found. Due to pooling different roles share the same connection and it's quite handy, because I can squeeze in a lot of users on a single connection with that approach (given they are mostly idle users).
And making the pg bg processes stay up as long as possible becomes such a handy feature as well...

**Comment by pkit** [comment]
> btw, I will note that I am a fan of zerovm :)

Cool, thanks :)
The idea I'm working on right now can be considered a continuation of that effort in a javascript land :)

**Comment by JerrySievert** [comment]
> The latter case is exactly the case I have found.

ok, I'd prefer PRs against the 3.0alpha branch (default), and if need be I can back port.

I don't suppose you have good experience with the inspector?  that's one larger thing that I haven't dived into fully: replacing the debugger that was removed after v8 3.14 with the newer inspector logic.  I have a branch that I've pulled in part of the d8 inspector, but haven't had the time/energy to revisit it and get it complete.

> The idea I'm working on right now can be considered a continuation of that effort in a javascript land :)

I'll be interested to see it!

**Comment by pkit** [comment]
I have already looked into inspector. Debugging with logs is not bad for now. But will need to implement it anyway later. 
I have also already done some async execution stuff (setTimeout and friends) although client->server->client promises still look scary...

**Comment by pkit** [comment]
BTW inspector requires websocket server and some js code. Therefore I would use node.js server as a proxy to postgres for that. I'm also using node.js as an http(websocket)->sql proxy (psql-like client) and auth proxy (oauth) for postgres.

**Comment by JerrySievert** [comment]
that inspector requires websockets and is no longer self-contained is one of the reasons i've held off for so long - pg can handle sockets cross-platform, but i really don't want to be trying to manage state for something that might be compiled by one or two people (it should never ship as a default for anything).
